### PR TITLE
implemented fix to dada2 dropping blank samples. 

### PR DESCRIPTION
**Pull request by dwthomas**
Fixes issue #136

### CODE REVIEW

**Review by thermokarst — COMMENTED** [review]
Thanks @dwthomas - this sounds like a good approach to me. Two points:

1. Should this behavior be put behind a parameter flag, like `--p-retain-all-samples`? If so, maybe we should make the existing behavior the default setting for that parameter, that way folks can opt _into_ this sample retention scheme. I am concerned that there might be downstream implications of retaining empty samples, and while the solution is simple (filter the empty samples out), it might be a stumbling block for new users. Thoughts?
2. This is missing unit tests. If you want a hand with writing some, just let me know, I am happy to lend a hand.

We are gearing up for the 2021.4 release, so if you want this to ship in this release, we only have another day to make that happen (sorry for the late response here) - just let me know how you want to proceed. Thanks!

**Comment by dwthomas** [comment]
Thanks @thermokarst 
1. As it stands dada2 will retain empty samples if they are not empty until the merging step of dada2, which to me seems like a bug, as you'd want all the empty samples or none.  I think a flag might be a good but separate idea as I agree there are probably some folks assuming there are no empty samples, as you say.
2. Some help navigating the unit tests would be useful!  I was having trouble finding where to put/get the input for a unit test.  I believe just adding another sample with no reads should demonstrate the behavior.

**Comment by thermokarst** [comment]
Thanks @dwthomas. So to make sure I'm following, would an alternative solution be to instead _filter_ all empty samples from the final feature table? If so, why didn't you take that approach here?

**Comment by dwthomas** [comment]
I think the two behaviors that make sense are:
1. All samples input to dada2 are retained, even with zero frequency.
2. Samples with zero frequency after dada2 are removed.
The current behavior retains some but not all zero frequency samples, which I think is a potential source of confusion.
I implemented the first because that's the behavior I want.  But my argument for that behavior as the default is that you can go from that to retaining only non-zero samples with a qiime feature-table filter-samples command, but I do not know a way to add back removed samples without using the Artifact API.

**Comment by thermokarst** [comment]
Awesome, thanks for explaining, that makes sense to me! Okay yeah, I'm leaning towards going with the functionality proposed here. In the future I think it would be great to add in a new flag that lets us toggle the empty-sample retention, wholesale, but I agree, that is a separate issue. I'll think about tests and get back to you. Would you mind if I push the tests up to your feature branch?

**Comment by dwthomas** [comment]
That works great for me.

**Comment by dwthomas** [comment]
@thermokarst I'm doing some work on the unit tests, I figured out where the data files are.

**Comment by dwthomas** [comment]
Hey @thermokarst I added unit tests and a toggle as discussed.  All the functionality is in the _denoise_helper function used by all single, paired and pyro so I modified the paired tests to include an empty sample and test both modes of the retain-all-samples toggle.

File: q2_dada2/_denoise.py
Comment (thermokarst) [inline q2_dada2/_denoise.py:145]: Probably best to comment this to explain that there are other sources of empty samples, not just the 6ish lines above this (otherwise, you might be tempted to stash everything in an if/else statement).

File: q2_dada2/tests/data/sample_seqs_paired/MANIFEST
Comment (thermokarst) [inline q2_dada2/tests/data/sample_seqs_paired/MANIFEST:2]: What happened with all of these files? It looks like maybe they all got renamed or something?

File: q2_dada2/tests/test_denoise.py
Comment (thermokarst) [inline q2_dada2/tests/test_denoise.py:27]: That wasn't necessary before, and I'm not sure what/why the changes in this PR prompted this change - did you run into a problem with unstable data during these tests?

**Review by thermokarst — CHANGES_REQUESTED** [review]

File: q2_dada2/_denoise.py
Comment (dwthomas) [reply q2_dada2/_denoise.py]: Great Point!

File: q2_dada2/_denoise.py
Comment (dwthomas) [reply q2_dada2/_denoise.py:140]: I get "TypeError: iteration over a 0-d array" when I pass a set to the Biom table constructor in the following line.  Which is why I did the conversion to list.  I agree it seems silly.

File: q2_dada2/tests/data/sample_seqs_paired/MANIFEST
Comment (dwthomas) [reply q2_dada2/tests/data/sample_seqs_paired/MANIFEST:2]: Yeah they did, I tried adding the "Empty" sample by modifying the Manifest file, but they were still being ignored.  So I imported them into qiime with 2021.2, then exported them, which worked but it seems to have renamed a lot of the files.

File: q2_dada2/_denoise.py
Comment (thermokarst) [reply q2_dada2/_denoise.py:140]: Oh sorry, I missed that below, I thought you were _only_ using this to get the total id count, my mistake!

File: q2_dada2/tests/test_denoise.py
Comment (dwthomas) [reply q2_dada2/tests/test_denoise.py:27]: Biom.table.concat sorts the opposite axis, see line 3391 here: https://github.com/biocore/biom-format/blob/master/biom/table.py, seems weird that it is not an option since it means there is not an equivalent to the additive identity of zero since even if there is nothing to concatenate you sort the opposite axis.

I chose to add the sorting prior to comparing:
1. I wanted to minimize changes in the _denoise code.
2. I did not want to modify the expected tables except adding a column for the new BLANK sample (to make it more clear)
3. This was easy to implement and I think clear what it is doing, maybe I should add a comment about why it became necessary?

File: q2_dada2/_denoise.py
Comment (dwthomas) [reply q2_dada2/_denoise.py:150]: Good idea!  I inserted one.

**Review by thermokarst — APPROVED** [review]

File: q2_dada2/_denoise.py
Comment (thermokarst) [inline q2_dada2/_denoise.py:88]:
```suggestion
    'retain_all_samples': _BOOL,
```

File: q2_dada2/_denoise.py
Comment (thermokarst) [inline q2_dada2/_denoise.py:446]:
```suggestion
        return _denoise_helper(biom_fp, track_fp, hashed_feature_ids)

```

**Comment by dwthomas** [comment]
Looks like there is just a new dada2 flavor that needed the toggle added to it as well.

**Comment by Oddant1** [comment]
Will this be ready to merge after the conflict is resolved?

**Comment by Oddant1** [comment]
@dwthomas are you still working on this PR?

**Comment by cherman2** [comment]
Hi @dwthomas! 
We are currently cleaning out our forum backlog! I fixed conflicts but it looks like there are still some errors! Do you have the bandwidth to take this pull request across the finish line? No pressure if not!

**Comment by dwthomas** [comment]
Hey @cherman2,
I unfortunately don't have the bandwidth at least for the time being.

**Comment by cherman2** [comment]
No problem! 
All these remaining issues look super superficial  (and I may have caused them when I fixed conflicts). I will get this passing and then pull down to test locally!

**Review by cherman2 — APPROVED** [review]
This looks good to me as long as CI passes! 👏 

@dwthomas, Thank you for all your work on this! This is awesome.

**Review by hagenjp — APPROVED** [review]
looks good to me! Thank you, @dwthomas

**Pull request merged**
